src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductEligibilityValidator.php

```diff
@@ -56,6 +63,11 @@ public function validate(mixed $value, Constraint $constraint): void
                     $constraint->message,
                     ['%productName%' => $orderItem->getProduct()->getName()],
                 );
+            } elseif (!$orderItem->getProduct()->hasChannel($channel)) {
+                $this->context->addViolation(
+                    $constraint->message,
+                    ['%productName%' => $orderItem->getProduct()->getName()],
+                );
```

Comment:
This should also be applied to the CoreBundle's validator.
We're also missing behat tests in ui and api contexts for this case.